cmd/soroban-rpc/internal/daemon/daemon.go

```diff
@@ -187,16 +192,54 @@ func MustNew(cfg config.LocalConfig) *Daemon {
 	}
 }
 
-func Run(cfg config.LocalConfig, endpoint string) {
+func Run(cfg config.LocalConfig, endpoint string, adminEndpoint string) {
 	d := MustNew(cfg)
-	supporthttp.Run(supporthttp.Config{
-		ListenAddr: endpoint,
-		Handler:    d,
-		OnStarting: func() {
-			d.logger.Infof("Starting Soroban JSON RPC server on %v", endpoint)
-		},
-		OnStopped: func() {
-			d.Close()
-		},
-	})
+
+	server := &http.Server{
+		Addr:        endpoint,
+		Handler:     d,
+		ReadTimeout: defaultReadTimeout,
+	}
+
+	d.logger.Infof("Starting Soroban JSON RPC server on %v", endpoint)
+
+	go func() {
+		if err := server.ListenAndServe(); !errors.Is(err, http.ErrServerClosed) {
+			// Error starting or closing listener:
+			d.logger.Fatalf("Soroban JSON RPC server encountered fatal error: %v", err)
+			os.Exit(1)
+		}
+	}()
+	var adminServer *http.Server
+	if adminEndpoint != "" {
+		// after importing net/http/pprof, debug endpoints are implicitly registered in the default serve mux
+		adminServer = &http.Server{Addr: adminEndpoint, Handler: http.DefaultServeMux}
+		go func() {
+			if err := adminServer.ListenAndServe(); !errors.Is(err, http.ErrServerClosed) {
+				d.logger.Errorf("Soroban admin server encountered fatal error: %v", err)
+			}
+		}()
+	}
+
+	// Shutdown gracefully when we receive an interrupt signal.
+	// First server.Shutdown closes all open listeners, then closes all idle connections.
+	// Finally, it waits a grace period (10s here) for connections to return to idle and then shut down.
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
+	<-signals
+
+	// Default supporthttp Shutdown grace period.
+	shutdownCtx, shutdownRelease := context.WithTimeout(context.Background(), defaultShutdownGracePeriod)
+	defer shutdownRelease()
+
+	if err := server.Shutdown(shutdownCtx); err != nil {
+		// Error from closing listeners, or context timeout:
+		d.logger.Errorf("Error during Soroban JSON RPC server Shutdown: %v", err)
+	}
+	d.Close()
+
+	if err := adminServer.Shutdown(shutdownCtx); err != nil {
```

Comment:
this would panic if `adminServer` is nil. I suggest that you'll add a defer function right after `go func() {` in line 217.

Reply:
It should, I actually prepared it for it, but forgot to check for nil. I don't think there is a need for defer